### PR TITLE
Add `MMDiT` for StableDiffusionV3

### PR DESCRIPTION
**Pull request by james77777778**
`MMDiT` is a complex model with many sublayers. I tried my best to simplify it but there might still be room for improvement.
Let me know if any updates are needed.

Numerics check:
https://colab.research.google.com/drive/1i__nshQpI8DPm0gy-vw0_GumRsd4NqMv?usp=sharing

Future works:
- [ ] Implement `ModelSamplingDiscreteFlow` and `CFGDenoiser`
- [ ] Implement `TextToImage` task class
- [ ] Implement StableDiffusionV3

@divyashreepathihalli @mattdangerw @SamanehSaadat

### CODE REVIEW

**Comment by mattdangerw** [comment]
This failure looks unrelated, and is just against `keras-nightly`, so I bet it's a new error that popped up over the past few days. Can ignore here.

**Review by mattdangerw — COMMENTED** [review]
Overall looks good! Just some minor style comments.

**Review by divyashreepathihalli — APPROVED** [review]
LGTM!

File: keras_nlp/src/models/stable_diffusion_v3/mmdit.py [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+        self.max_period = float(max_period)
+        self.half_frequency_dim = self.frequency_dim // 2
+
+        self.mlp = models.Sequential(
+            [
+                layers.Dense(
+                    embedding_dim,
+                    activation="silu",
+                    dtype=self.dtype_policy,
+                    name="dense0",
```
[the remark this answers is not in the document]
Comment (james77777778) [reply keras_nlp/src/models/stable_diffusion_v3/mmdit.py]: I removed the `name` from sublayers and instead added a `name` for the entire `Sequential`

File: keras_nlp/src/models/stable_diffusion_v3/mmdit.py [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+        config.update(
+            {
+                "patch_size": self.patch_size,
+                "hidden_dim": self.hidden_dim,
+            }
+        )
+        return config
+
+
+class PositionEmbedding(layers.Layer):
```
[the remark this answers is not in the document]
Comment (james77777778) [reply keras_nlp/src/models/stable_diffusion_v3/mmdit.py]: I implemented a new layer called `AdjustablePositionEmbedding` which is a subclass of `PositionEmbedding`.

It allows us to adjust the shape of embedding and slice it based on the input height and width, which `PositionEmbedding` could not accommodate.

File: keras_nlp/src/models/stable_diffusion_v3/mmdit.py [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+                num_heads,
+                hidden_dim,
+                mlp_ratio,
+                use_context_projection=not (i == depth - 1),
+                dtype=dtype,
+                name=f"joint_block{i}",
+            )
+            for i in range(depth)
+        ]
+        self.final_layer = OutputLayer(
```
[the remark this answers is not in the document]
Comment (james77777778) [reply keras_nlp/src/models/stable_diffusion_v3/mmdit.py]: Changed to `output_layer`

File: keras_nlp/src/models/stable_diffusion_v3/mmdit.py [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+        # Embeddings.
+        x = self.patch_embedding(latent_inputs)
+        cropped_position_embedding = self._get_cropped_position_embedding(
+            x, patch_size, image_size, position_size
+        )
+        x = layers.Add(dtype=dtype)([x, cropped_position_embedding])
+        context = self.context_embedding(context_inputs)
+        pooled_projection = self.vector_embedding(pooled_projection_inputs)
+        timestep_embedding = self.timestep_embedding(timestep_inputs)
+        timestep_embedding = layers.Add(dtype=dtype)(
```
[the remark this answers is not in the document]
Comment (james77777778) [reply keras_nlp/src/models/stable_diffusion_v3/mmdit.py]: Thanks for pointing it out. Fixed.

**Review by mattdangerw — APPROVED** [review]
LGTM!

**Pull request merged**